// @echristian/edge-tts/dist/main.js
// src/lib/constants.ts
var API_CONFIG = {
  baseUrl: "https://speech.platform.bing.com",
  headers: {
    Authority: "speech.platform.bing.com",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 Edg/130.0.0.0",
    Accept: "*/*",
    "Accept-Language": "en-US,en;q=0.9",
    "Accept-Encoding": "gzip, deflate, br",
    "Sec-CH-UA": '" Not;A Brand";v="99", "Microsoft Edge";v="130", "Chromium";v="130"',
    "Sec-CH-UA-Mobile": "?0",
    "Sec-Fetch-Site": "none",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Dest": "empty"
  }
};
var DEFAULT_OPTIONS = {
  voice: "en-US-AvaNeural",
  language: "en-US",
  outputFormat: "audio-24khz-96kbitrate-mono-mp3",
  rate: "default",
  pitch: "default",
  volume: "default",
  subtitle: {
    splitBy: "word",
    wordsPerCue: 10,
    durationPerCue: 5e3
  }
};
var WEBSOCKET_CONFIG = {
  baseUrl: "wss://speech.platform.bing.com",
  path: "/consumer/speech/synthesize/readaloud/edge/v1",
  trustedClientToken: "6A5AA1D4EAFF4E9FB37E23D68491D6F4"
};
var AUDIO_CONSTANTS = {
  TIME_FACTOR: 1e4
};

// src/get-voices.ts
async function getVoices() {
  const url = `${API_CONFIG.baseUrl}/consumer/speech/synthesize/readaloud/voices/list?trustedclienttoken=${WEBSOCKET_CONFIG.trustedClientToken}`;
  try {
    const response = await fetch(url, {
      headers: API_CONFIG.headers
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return await response.json();
  } catch (error) {
    console.error("Error fetching voices:", error);
    throw error;
  }
}

// src/lib/connection.ts
function createUrl() {
  const searchParams = new URLSearchParams({
    TrustedClientToken: WEBSOCKET_CONFIG.trustedClientToken
  });
  const url = new URL(WEBSOCKET_CONFIG.path, WEBSOCKET_CONFIG.baseUrl);
  url.search = searchParams.toString();
  return url;
}
function createInitialMessage(outputFormat) {
  return `Content-Type:application/json; charset=utf-8\r
Path:speech.config\r
\r

    {
      "context": {
        "synthesis": {
          "audio": {
            "metadataoptions": {
              "sentenceBoundaryEnabled": "false",
              "wordBoundaryEnabled": "true"
            },
            "outputFormat": "${outputFormat}"
          }
        }
      }
    }`;
}
function createSocket(outputFormat) {
  if (!outputFormat) throw new Error("Output format is required");
  const ws = new WebSocket(createUrl().toString());
  const initialMessage = createInitialMessage(outputFormat);
  const { promise, resolve } = Promise.withResolvers();
  ws.addEventListener("open", () => {
    ws.send(initialMessage);
    resolve(ws);
  });
  return promise;
}

// src/subtitle.ts
var convertToMs = (duration) => Math.floor(duration / AUDIO_CONSTANTS.TIME_FACTOR);
function parseSubtitle({
  splitBy,
  wordsPerCue,
  durationPerCue,
  metadata
}) {
  const simplified = metadata.map((meta) => ({
    text: meta.Metadata[0].Data.text.Text,
    offset: convertToMs(meta.Metadata[0].Data.Offset),
    duration: convertToMs(meta.Metadata[0].Data.Duration)
  }));
  if (splitBy === "duration") {
    if (durationPerCue === void 0)
      throw new Error(
        "durationPerCue option must be provided when splitting by duration"
      );
    return simplified.reduce(
      (prev, curr, index) => {
        if ((prev.at(-1)?.duration ?? 0) + curr.duration > durationPerCue || index === 0) {
          prev.push({
            text: curr.text,
            start: curr.offset,
            duration: curr.duration,
            end: curr.offset + curr.duration
          });
        } else {
          prev[prev.length - 1].end = curr.offset + curr.duration;
          prev[prev.length - 1].text += ` ${curr.text}`;
          prev[prev.length - 1].duration = prev[prev.length - 1].end - prev[prev.length - 1].start;
        }
        return prev;
      },
      []
    );
  }
  if (splitBy === "word") {
    if (wordsPerCue === void 0)
      throw new Error(
        "wordsPerCue option must be provided when splitting by word"
      );
    return simplified.reduce(
      (prev, curr, index) => {
        const wordCount = prev.at(-1)?.text.split(/\s+/).length ?? 0;
        if (wordCount >= wordsPerCue || index === 0) {
          prev.push({
            text: curr.text,
            start: curr.offset,
            duration: curr.duration,
            end: curr.offset + curr.duration
          });
        } else {
          prev[prev.length - 1].end = curr.offset + curr.duration;
          prev[prev.length - 1].text += ` ${curr.text}`;
          prev[prev.length - 1].duration = prev[prev.length - 1].end - prev[prev.length - 1].start;
        }
        return prev;
      },
      []
    );
  }
  throw new Error("Invalid splitBy option");
}

// src/lib/audio-processor.ts
var AUDIO_PATH_SEPARATOR = "Path:audio\r\n";
async function processAudioChunk(chunk) {
  const bytes = new Uint8Array(await chunk.arrayBuffer());
  const binaryString = new TextDecoder().decode(bytes);
  const index = binaryString.indexOf(AUDIO_PATH_SEPARATOR) + AUDIO_PATH_SEPARATOR.length;
  return bytes.subarray(index);
}
async function processAudioChunks(chunks) {
  const processed = [];
  for (const chunk of chunks) {
    processed.push(await processAudioChunk(chunk));
  }
  return processed;
}
function toBlobLike(data) {
  if (data instanceof Blob) return data;
  return new Blob([data]);
}

// src/lib/metadata-processor.ts
function parseMetadataMessage(message) {
  const hasMetadata = message.includes("Path:audio.metadata");
  if (!hasMetadata) return void 0;
  const jsonString = message.split("Path:audio.metadata")[1].trim();
  return JSON.parse(jsonString);
}

// src/lib/socket-handler.ts
function createInitialState() {
  return {
    audioChunks: [],
    subtitleChunks: []
  };
}
function handleStringMessage(message, state, socket) {
  if (message.includes("Path:turn.end")) {
    socket.close();
    return;
  }
  const metadata = parseMetadataMessage(message);
  if (metadata !== void 0) {
    state.subtitleChunks.push(metadata);
  }
}
function handleBinaryMessage(data, state) {
  state.audioChunks.push(toBlobLike(data));
}
function handleTTSConnection(socket, options) {
  const state = createInitialState();
  const { promise, resolve, reject } = Promise.withResolvers();
  socket.addEventListener("error", reject);
  socket.addEventListener("close", async () => {
    const processedChunks = await processAudioChunks(state.audioChunks);
    resolve({
      audio: new Blob(processedChunks),
      subtitle: parseSubtitle({ metadata: state.subtitleChunks, ...options })
    });
  });
  socket.addEventListener(
    "message",
    ({ data }) => {
      if (typeof data === "string") {
        handleStringMessage(data, state, socket);
      } else {
        handleBinaryMessage(data, state);
      }
    }
  );
  return promise;
}

// src/lib/ssml.ts
function createSSMLString(params) {
  return `
  X-RequestId:${params.requestId}\r

  Content-Type:application/ssml+xml\r

  Path:ssml\r
\r


  <speak version="1.0" xmlns="http://www.w3.org/2001/10/synthesis" xmlns:mstts="https://www.w3.org/2001/mstts" xml:lang="${params.language}">
    <voice name="${params.voice}">
      <prosody rate="${params.rate}" pitch="${params.pitch}" volume="${params.volume}">
        ${params.text}
      </prosody>
    </voice>
  </speak>
  `;
}

// src/synthesize.ts
async function synthesize(options) {
  const voice = options.voice ?? DEFAULT_OPTIONS.voice;
  const language = options.language ?? DEFAULT_OPTIONS.language;
  const outputFormat = options.outputFormat ?? DEFAULT_OPTIONS.outputFormat;
  const rate = options.rate ?? DEFAULT_OPTIONS.rate;
  const pitch = options.pitch ?? DEFAULT_OPTIONS.pitch;
  const volume = options.volume ?? DEFAULT_OPTIONS.volume;
  const subtitle = {
    splitBy: options.subtitle?.splitBy ?? DEFAULT_OPTIONS.subtitle.splitBy,
    wordsPerCue: options.subtitle?.wordsPerCue ?? DEFAULT_OPTIONS.subtitle.wordsPerCue,
    durationPerCue: options.subtitle?.durationPerCue ?? DEFAULT_OPTIONS.subtitle.durationPerCue
  };
  const socket = await createSocket(outputFormat);
  const requestId = globalThis.crypto.randomUUID();
  const requestString = createSSMLString({
    requestId,
    text: options.text,
    voice,
    language,
    rate,
    pitch,
    volume
  });
  const result = handleTTSConnection(socket, subtitle);
  socket.send(requestString);
  return result;
}

// src/synthesize-stream.ts
async function* synthesizeStream(options) {
  const voice = options.voice ?? DEFAULT_OPTIONS.voice;
  const language = options.language ?? DEFAULT_OPTIONS.language;
  const outputFormat = options.outputFormat ?? DEFAULT_OPTIONS.outputFormat;
  const rate = options.rate ?? DEFAULT_OPTIONS.rate;
  const pitch = options.pitch ?? DEFAULT_OPTIONS.pitch;
  const volume = options.volume ?? DEFAULT_OPTIONS.volume;
  const socket = await createSocket(outputFormat);
  const requestId = globalThis.crypto.randomUUID();
  const requestString = createSSMLString({
    requestId,
    text: options.text,
    voice,
    language,
    rate,
    pitch,
    volume
  });
  const { promise: done, resolve, reject } = Promise.withResolvers();
  const queue = [];
  let isDone = false;
  let error = null;
  socket.addEventListener("error", (err) => {
    error = err instanceof Error ? err : new Error("WebSocket error");
    reject(error);
  });
  socket.addEventListener(
    "message",
    ({ data }) => {
      if (typeof data === "string") {
        if (data.includes("Path:turn.end")) {
          socket.close();
          isDone = true;
          resolve();
        }
      } else {
        queue.push(toBlobLike(data));
      }
    }
  );
  socket.send(requestString);
  try {
    while (!isDone || queue.length > 0) {
      if (error) throw error;
      if (queue.length > 0) {
        const chunk = queue.shift();
        const processed = await processAudioChunk(chunk);
        yield processed;
      } else {
        await new Promise((resolve2) => setTimeout(resolve2, 10));
      }
    }
    await done;
  } finally {
    if (socket.readyState === WebSocket.OPEN) {
      socket.close();
    }
  }
}
export {
  getVoices,
  synthesize,
  synthesizeStream
};
